addons/account/models/account_journal.py

```diff
@@ -403,9 +403,26 @@ def unlink(self):
     @api.returns('self', lambda value: value.id)
     def copy(self, default=None):
         default = dict(default or {})
+
+        # Find a unique code for the copied journal
+        read_codes = self.env['account.journal'].read_group([('company_id', '=', self.company_id.id)], [], ['code'])
+        all_journal_codes = {code_data['code'] for code_data in read_codes}
+
+        copy_code = self.code
+        code_prefix = re.sub('\d+', '', self.code).strip()
+        counter = 1
+        while counter <= len(all_journal_codes) and  copy_code in all_journal_codes:
+            copy_code = ("%s%s" % (code_prefix, counter))[-self._fields['code'].size:]
```

Comment:
not sure left stripping the code is the best idea: keeping the leftmost letters is more meaningful in many languages.

ex: journal copied with code SALES => ALES1 instead of SALE1 ?